File: ios/Classes/Extensions/MediaInfoExtensions.swift
//
//  MediaInfoExtensions.swift
//  google_cast
//
//  Created by LUIZ FELIPE ALVES LIMA on 27/06/22.
//

import Foundation
import GoogleCast

extension GCKMediaInformation{
    
    
    static func fromMap(_ arguments: Dictionary<String, Any> ) -> GCKMediaInformation? {
        
        let contentID = arguments["contentID"] as! String
        let streamType = GCKMediaStreamType.unknown
        guard let  contentUrl = URL.init(string: arguments["contentURL"] as! String) else { return nil}
        let builder =  GCKMediaInformationBuilder.init()
        builder.contentID  = contentID
        builder.streamType = streamType
        builder.contentURL = contentUrl
        builder.customData = arguments["customData"]
      
        
    
        if let tracksDict = arguments["tracks"] as? [Dictionary<String, Any>] {
    builder.mediaTracks = tracksDict.map{
                dict in
                GCKMediaTrack.fromMap(dict)
            }
        }
      
        if let metadataDict = arguments["metadata"] {
        builder.metadata = GCKMediaMetadata.fromMap(metadataDict as! Dictionary<String, Any>)
        }
      
        let buildedMediaInfo = builder.build()
        
        return buildedMediaInfo
    
    }
    
    
    
    
    func toMap() -> Dictionary<String, Any> {
        var dict = Dictionary<String, Any>()
        dict["contentID"] = self.contentID
        dict["contentType"] = self.contentType
        dict["streamType"] = self.streamType.rawValue
        dict["contentURL"] = self.contentURL?.absoluteString
        dict["duration"] = self.streamDuration
        dict["metadata"] = self.metadata?.toMap()
        if(self.mediaTracks != nil){
            dict["tracks"] = self.mediaTracks!.map{
                track in
                track.toMap()
                
            }
        }
        
      
        
        
        
        return dict
    }
    
}

File: android/src/main/kotlin/com/felnanuke/google_cast/extensions/MediaInfoExtensions.kt
package com.felnanuke.google_cast.extensions

import com.google.android.gms.cast.MediaInfo
import com.google.android.gms.cast.MediaMetadata
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import org.json.JSONObject

/**
 * Utility class for building [MediaInfo] objects for Google Cast from map representations.
 */
class GoogleCastMediaInfo {
    companion object {
        /**
         * Creates a [MediaInfo] instance from a [Map] representation.
         *
         * @param map A [Map] containing media information. Expected keys:
         *  - "contentID": [String] (required) - The unique content ID.
         *  - "contentURL": [String]? - The URL of the content.
         *  - "contentType": [String]? - The MIME type of the content.
         *  - "tracks": [List]<[Map]<[String], Any?>>? - List of media track maps.
         *  - "metadata": [Map]<[String], Any?>? - Metadata map.
         *  - "customData": [Map]<[String], Any?>? - Custom data map.
         * @return A [MediaInfo] object if the required fields are present, otherwise null.
         */
        fun fromMap(map: Map<String, Any?>): MediaInfo? {
            val contentId = map["contentID"] as String
            val contentUrl = map["contentURL"] as String?
            val contentType = map["contentType"] as String?
            val tracks = GoogleCastMediaTrackBuilder.listFromMap(
                map["tracks"] as List<Map<String, Any?>>? ?: listOf()
            )
            val metaData =
                GoogleCastMetadataBuilder.fromMap(map["metadata"] as Map<String, Any?>? ?: mapOf())
            val builder = MediaInfo.Builder(contentId)
            if (contentUrl != null)
                builder.setContentUrl(contentUrl)
            if (contentType != null)
                builder.setContentType(contentType)
            if (tracks.isNotEmpty())
                builder.setMediaTracks(tracks)

            if (metaData != null)
                builder.setMetadata(metaData)
            var customData = map["customData"] as Map<String, Any?>?

            if (customData != null) {
                builder.setCustomData(JSONObject(customData))
            }

            return builder.build()
        }
    }
}

//fun MediaInfo.toMap(): Map<String, Any?> {
//    var map = mutableMapOf<String, Any?>()
//    map["contentID"] = contentId
//    map["contentURL"] = contentUrl
//    map["contentType"] = contentType
//    map["streamType"] = streamType
//    map["streamDuration"] = streamDuration
//    map["tracks"] = mediaTracks?.map {
//        it.toMap()
//    }
//    map["metadata"] = null
//    return map
//
//}